### PR TITLE
Issue388 peak demand - electricity, gas, dist heat

### PR DESCRIPTION
**Pull request by dhblum**
This is for #388 and implements a peak demand core KPI, for electricity only.

Still to do:
- [x] Update design docs
- [x] Review

### CODE REVIEW

**Comment by dhblum** [comment]
@JavierArroyoBastida I've implemented peak demand for electricity.  I plan to also implement for gas, but can you review this for now to see what you think?  Gas will follow what's done here.  I know in WP1.2 we talked about implementing for each energy vector, but I'm a bit unsure of the worthwhileness of this since we need to implement a new kpi for each vector and vectors outside of gas and electricity aren't really used in any of the cases yet.  Could be discussed more in next WP1.2 meeting.

File: kpis/kpi_calculator.py [hunk 1]
```diff
@@ -345,6 +358,53 @@ def get_energy(self, plot=False, plot_by_source=False):
...
+            df_pow_data_all = pd.DataFrame(index=tim_data)
+            # Calculate peak electricity
+            # [returns KW/m^2 - assumes power measured in Watts]
+            for source in self.sources_pele:
+                for signal in self.case.kpi_json[source]:
+                    pow_data = np.array(self._get_data_from_last_index(signal,self.i_last_pele))
+                    df_pow_data = pd.DataFrame(index=tim_data, data=pow_data, columns=[signal])
+                    df_pow_data_all = pd.concat([df_pow_data_all, df_pow_data], axis=1)
+            df_pow_data_all.index = pd.TimedeltaIndex(df_pow_data_all.index, unit='s')
+            df_pow_data_all = df_pow_data_all.resample('15T').mean()/self.case._get_area()/1000.
```
Comment (JavierArroyoBastida) [inline kpis/kpi_calculator.py:392]: Are we sure we want to take the mean value here? Taking the mean may average out the peak values but I think we're interested in the instantaneous peak. If so, interpolation may be required.

**Review by JavierArroyoBastida — COMMENTED** [review]
@dhblum I've suggested a few comments/reflections. Please let me know your thoughts.

File: kpis/kpi_calculator.py [hunk 2]
```diff
@@ -345,6 +358,53 @@ def get_energy(self, plot=False, plot_by_source=False):
...
+            df_pow_data_all['peak_demand'] = df_pow_data_all.sum(axis=1)
```
[the remark this answers is not in the document]
Comment (dhblum) [reply kpis/kpi_calculator.py:393]: Agreed, I will make a fix.

File: kpis/kpi_calculator.py [hunk 3]
```diff
@@ -345,6 +358,53 @@ def get_energy(self, plot=False, plot_by_source=False):
...
+            peak = df_pow_data_all['peak_demand'].max()
+            self.pele_tot = peak
+            # Find contributions to peak by each signal
+            i = df_pow_data_all['peak_demand'].idxmax()
```
[the remark this answers is not in the document]
Comment (dhblum) [reply kpis/kpi_calculator.py:397]: I think you're right on your second suggestion so as not to have to find the max twice, which I suppose will be more efficient.  I'll test your suggestion.

File: kpis/kpi_calculator.py (hunk 1 above)
Comment (dhblum) [reply kpis/kpi_calculator.py:392]: Let's discuss on the issues page (https://github.com/ibpsa/project1-boptest/issues/388).

File: kpis/kpi_calculator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhblum) [reply kpis/kpi_calculator.py]: Addressed in https://github.com/ibpsa/project1-boptest/pull/394/commits/1d2410e9cfcbecc890f9cfd374ca716e4dc7bce2.

File: kpis/kpi_calculator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhblum) [reply kpis/kpi_calculator.py]: Addressed in https://github.com/ibpsa/project1-boptest/pull/394/commits/9b75ed9c2e41079325bb18b5dec911d5434a1ddc.  Worked just fine.

**Comment by dhblum** [comment]
@JavierArroyoBastida This is getting pretty close and I think is finished for electricity pending tests pass and after our discussion at the last WP1.2 meeting regarding using 15-min average.  I plan to duplicate for gas and district heating.  Let me know any further comments if you have them.

File: docs/DesReqGui/source/core_kpi.rst [hunk 4]
```diff
@@ -51,6 +51,23 @@ Energy use in a given period of time
...
+   when accounting for the sum of all electrical equiipment present in the test
+   case HVAC system.
+   The mathematical formulation for this KPI is the following:
+
+   .. math:: P_e(t_0, t_f) = \frac{\max_{15}[\sum_{i\in \epsilon} P_i(t)]}{A}
+
+   Where :math:`P_e(t_0, t_f)` is the peak electricity demand from the
+   initial time :math:`t_0` up to the final time :math:`t_f`; :math:`\epsilon` denotes
+   the set of equipment in the system with an associated energy use of
+   electricity; :math:`P_i` is the instantaneous power used by the
```
Comment (JavierArroyoBastida) [inline docs/DesReqGui/source/core_kpi.rst:67]: It is a bit strange that we talk here about a generic energy vector  `i` when in this case there is only one energy vector which is electricity. I would reformulate using just `P`, the instantaneous electricity use.

**Review by JavierArroyoBastida — CHANGES_REQUESTED** [review]
Thanks, @dhblum again for moving this forward. I just have a couple of suggestions for the documentation of the new KPI.

File: docs/DesReqGui/source/core_kpi.rst (hunk 4 above)
Comment (dhblum) [reply docs/DesReqGui/source/core_kpi.rst:67]: How about my suggestion in https://github.com/ibpsa/project1-boptest/pull/394/commits/b03ed26fbc7df72babe727260a0f83cce6afc7ea to just make the set of all equipment ``I``?  And I'd suggest just redefining ``I`` locally for Gas and District peaks.

**Comment by dhblum** [comment]
@JavierArroyoBastida thanks for the review.  I think I will replicate for Gas and District.

File: docs/DesReqGui/source/core_kpi.rst (hunk 4 above)
Comment (JavierArroyoBastida) [reply docs/DesReqGui/source/core_kpi.rst:67]: Looking good. Thanks for taking care of it!

**Comment by dhblum** [comment]
@JavierArroyoBastida Gas and district heating peaks added, tests are passing.  I think it's ready to merge.  Take one more look with the new additions?

File: testing/references/bestest_air/kpis_initialize_initial.csv [hunk 5]
```diff
@@ -1,7 +1,10 @@
 keys,value
+tdis_tot,0.0
+idis_tot,0.0
 ener_tot,0.0
+cost_tot,0.0
 emis_tot,0.0
-idis_tot,0.0
-tdis_tot,0.0
+pele_tot,4.647544789502356e-36
```
Comment (JavierArroyoBastida) [inline testing/references/bestest_air/kpis_initialize_initial.csv:7]: Any idea why is this happening?

**Review by JavierArroyoBastida — APPROVED** [review]
Looking good @dhblum! I've added a minor suggestion and a small question.

File: testing/references/bestest_air/kpis_initialize_initial.csv (hunk 5 above)
Comment (JavierArroyoBastida) [reply testing/references/bestest_air/kpis_initialize_initial.csv:7]: I was imagining something like that, but this explains better. Thanks! I think we can live with such a small numerical diffusion.

File: docs/DesReqGui/source/core_kpi.rst [hunk 6]
```diff
@@ -51,6 +51,57 @@ Energy use in a given period of time
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+   This KPI is the measure of the site HVAC peak district heating use in :math:`kW/m^2`
+   when accounting for the sum of all district heating equipment present in the test
+   case HVAC system.
+   The mathematical formulation for this KPI is the following:
+
+   .. math:: P_{DH}(t_0, t_f) = \frac{\max_{15}[\sum_{i\in I} P_i(t)]}{A}
+
+   Where :math:`P_{Dh}(t_0, t_f)` is the peak district heating demand from the
```
Comment (JavierArroyoBastida) [inline docs/DesReqGui/source/core_kpi.rst:98]:
```suggestion
   Where :math:`P_{DH}(t_0, t_f)` is the peak district heating demand from the
```
Sorry I didn't notice the capitalization in my previous suggestion. Just to be coherent with the symbols used in the equation above.

**Review by JavierArroyoBastida — APPROVED** [review]

File: testing/references/bestest_air/kpis_initialize_initial.csv (hunk 5 above)
Comment (dhblum) [reply testing/references/bestest_air/kpis_initialize_initial.csv:7]: Yes, I think for now it is fine.

**Comment by dhblum** [comment]
Thank you @JavierArroyoBastida !!

**Pull request merged**